### PR TITLE
feat: add custom rate limits for organisations

### PR DESCRIPTION
**Pull request by maxdeichmann**

### CODE REVIEW

File: web/src/features/public-api/server/RateLimitService.ts [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+        { resource: "public-api-metrics", points: 100, duration: 60 },
+      ];
+      break;
+    case "self-hosted:enterprise":
+      planConfig = [
+        { resource: "ingestion", points: 10000, duration: 60 },
+        { resource: "prompts", points: null, duration: null },
+        { resource: "public-api", points: 20000, duration: 60 },
+        { resource: "public-api-metrics", points: 200, duration: 60 },
+      ];
```
Comment (maxdeichmann) [inline web/src/features/public-api/server/RateLimitService.ts:65]: fix the plan values

File: web/src/features/public-api/server/RateLimitService.ts [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+      planConfig = [
+        { resource: "ingestion", points: 2000, duration: 60 },
+        { resource: "prompts", points: null, duration: null },
+        { resource: "public-api", points: 5000, duration: 60 },
+        { resource: "public-api-metrics", points: 50, duration: 60 },
+      ];
+      break;
+    case "cloud:team":
+      planConfig = [
+        { resource: "ingestion", points: 5000, duration: 60 },
```
Comment (maxdeichmann) [inline web/src/features/public-api/server/RateLimitService.ts:53]: rename to durationInMin

File: web/src/features/public-api/server/RateLimitService.ts [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+        { resource: "public-api-metrics", points: 200, duration: 60 },
+      ];
+      break;
+    default:
+      // typescript type error if we don't handle all plans in the switch
+      const exhaustiveCheck: never = plan;
+      throw new Error(`Unhandled plan case: ${exhaustiveCheck}`);
+  }
+
+  const config = planConfig.find((config) => config.resource === resource);
```
Comment (maxdeichmann) [inline web/src/features/public-api/server/RateLimitService.ts:73]: apply the same switch statement

File: web/src/features/public-api/server/apiAuth.ts [hunk 4]
```diff
@@ -128,21 +157,34 @@ export class ApiAuthService {
...
         }
 
-        addUserToSpan({ projectId });
+        addUserToSpan({ projectId: finalApiKey.projectId });
+
+        const plan = finalApiKey.plan;
+
+        if (!isPlan(plan)) {
+          console.error("Invalid plan type for key", finalApiKey.plan);
+          throw new Error("Invalid credentials");
```
Comment (maxdeichmann) [inline web/src/features/public-api/server/apiAuth.ts:177]: Improve error message

File: web/src/features/public-api/server/RateLimitService.ts [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+    this.res = res;
+  }
+
+  isRateLimited() {
+    return this.res ? this.res.remainingPoints < 1 : false;
+  }
+
+  sendRestResponseIfLimited(nextResponse: NextApiResponse) {
+    if (!this.res || !this.isRateLimited()) {
+      return;
```
Comment (marcklingen) [inline web/src/features/public-api/server/RateLimitService.ts:139]: this should be an exception and must not happen, track this

File: web/src/features/public-api/server/apiAuth.ts [hunk 6]
```diff
@@ -154,17 +196,25 @@ export class ApiAuthService {
...
+          : undefined;
+
         return {
           validKey: true,
           scope: {
             projectId: dbKey.projectId,
             accessLevel: "scores",
+            orgId: dbKey.project.organization.id,
+            plan: getOrganizationPlan(cloudConfig),
+            rateLimits: cloudConfig?.rateLimits ?? [],
```
Comment (marcklingen) [inline web/src/features/public-api/server/apiAuth.ts:210]: how do the right rate limits get merged onto this scope?

File: packages/shared/src/interfaces/cloudConfigSchema.ts [hunk 7]
```diff
@@ -13,5 +14,9 @@ export const CloudConfigSchema = z.object({
       activeProductId: z.string().optional(),
     })
     .optional(),
+
+  // custom rate limits for an organization
+  rateLimits: CloudConfigRateLimit.nullable(),
```
Comment (marcklingen) [inline packages/shared/src/interfaces/cloudConfigSchema.ts:19]: rateLimits -> rateLimitOverrides

File: web/src/features/public-api/server/RateLimitService.ts [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+
+  constructor(redis: Redis | null) {
+    this.redis = redis;
+  }
+
+  async rateLimitRequest(
+    scope: ApiAccessScope,
+    resource: z.infer<typeof RateLimitResource>,
+  ) {
+    // if cloud config is not present, we don't apply rate limits and just return
```
Comment (marcklingen) [inline web/src/features/public-api/server/RateLimitService.ts:34]: we already include this check on plan level, necesary to add this again?

File: web/src/features/public-api/server/RateLimitService.ts [hunk 9]
```diff
@@ -0,0 +1,216 @@
...
+  res: NextApiResponse,
+  rateLimitRes: RateLimitResult,
+) => {
+  const httpHeader = createHttpHeaderFromRateLimit(rateLimitRes);
+
+  for (const [header, value] of Object.entries(httpHeader)) {
+    res.setHeader(header, value);
+  }
+
+  res.status(429).end();
```
Comment (marcklingen) [inline web/src/features/public-api/server/RateLimitService.ts:155]: shall we add a human readable response in addition to status and headers? e.g. simple "429 - rate limit exceeded"

File: web/src/features/public-api/server/RateLimitService.ts [hunk 10]
```diff
@@ -0,0 +1,216 @@
...
+    scope: ApiAccessScope,
+    resource: z.infer<typeof RateLimitResource>,
+  ) {
+    const planBasedConfig = getRateLimitConfig(scope.plan, resource);
+
+    const customConfig = scope.rateLimits?.find(
+      (config) => config.resource === resource,
+    );
+
+    const effectiveConfig = customConfig || planBasedConfig;
```
Comment (marcklingen) [inline web/src/features/public-api/server/RateLimitService.ts:57]: nit" i'd suggest ot move this into getRateLimitConfig as we can then reuse that function to render rate limits in the UI (iterating over all rate limits resources, can be in trpc route) while keeping the logic of overrides in a central function

**Review by marcklingen — COMMENTED** [review]
looks good to me, added a few comments but nothing major

**Pull request merged**